### PR TITLE
[virtctl] [cleancode] split the functionality of the VNC run function

### PR DESCRIPTION
**Pull request by enp0s3**
The function is being overloaded, therefore decouple
the local proxy creation and the VNC stream creation
into a separate receiver functions.

Signed-off-by: Igor Bezukh <ibezukh@redhat.com>

**What this PR does / why we need it**:
Code readability. Make the code easy for unit testing

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/virtctl/vnc/vnc.go [hunk 1]
```diff
@@ -84,37 +84,51 @@ type VNC struct {
...
 
-func (o *VNC) Run(cmd *cobra.Command, args []string) error {
-	namespace, _, err := o.clientConfig.Namespace()
+func (o *VNC) setupLocalProxy(vmiName string) (*net.TCPListener, error) {
+	lnAddr, err := net.ResolveTCPAddr("tcp", fmt.Sprintf("127.0.0.1:%d", customPort))
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("Can't resolve the address: %s", err.Error())
+	}
+	// The local tcp server is used to proxy the podExec websock connection to vnc client
```
Comment (dankenigsberg) [inline pkg/virtctl/vnc/vnc.go:92]: I believe that this comment explains the purpose of the function, it should probably be placed above the signature?
Maybe you can rename the function so the comment is not needed at all?

File: pkg/virtctl/vnc/vnc.go [hunk 2]
```diff
@@ -84,37 +84,51 @@ type VNC struct {
...
+	ln, err := net.ListenTCP("tcp", lnAddr)
+	if err != nil {
+		return nil, fmt.Errorf("Can't listen on unix socket: %s", err.Error())
 	}
 
-	vmi := args[0]
+	return ln, nil
```
Comment (maiqueb) [inline pkg/virtctl/vnc/vnc.go:98]: I wonder if it isn't simpler to just flat out `return net.ListenTCP("tcp", lnAddr)` . 

Do you think the error string you're adding is important enough ?

**Review by maiqueb — COMMENTED** [review]
Just a simple question, on this preliminary review.

File: pkg/virtctl/vnc/vnc.go [hunk 3]
```diff
@@ -84,37 +84,51 @@ type VNC struct {
 	clientConfig clientcmd.ClientConfig
 }
 
-func (o *VNC) Run(cmd *cobra.Command, args []string) error {
-	namespace, _, err := o.clientConfig.Namespace()
+func (o *VNC) setupLocalProxy(vmiName string) (*net.TCPListener, error) {
```
Comment (ezrasilvera) [inline pkg/virtctl/vnc/vnc.go:87]: Nit: I think a createXXX  would be better name here as you are actually creating a new listener here.

File: pkg/virtctl/vnc/vnc.go [hunk 4]
```diff
@@ -84,37 +84,51 @@ type VNC struct {
...
+	if err != nil {
+		return nil, fmt.Errorf("Can't listen on unix socket: %s", err.Error())
 	}
 
-	vmi := args[0]
+	return ln, nil
+}
 
-	virtCli, err := kubecli.GetKubevirtClientFromClientConfig(o.clientConfig)
+func (o *VNC) createVmiVncHandler(vmiName string) (kubecli.StreamInterface, error) {
```
Comment (ezrasilvera) [inline pkg/virtctl/vnc/vnc.go:101]: Maybe getVmiVNCStream   ?

File: pkg/virtctl/vnc/vnc.go (hunk 1 above)
Comment (enp0s3) [reply pkg/virtctl/vnc/vnc.go]: @dankenigsberg You are correct. Applied more informative function name.

File: pkg/virtctl/vnc/vnc.go (hunk 2 above)
Comment (enp0s3) [reply pkg/virtctl/vnc/vnc.go:98]: @maiqueb I would prefer to preserve the error message, I see that TCP listener errors are not so self explanatory so I think this extra info may help

File: pkg/virtctl/vnc/vnc.go (hunk 3 above)
Comment (enp0s3) [reply pkg/virtctl/vnc/vnc.go]: @ezrasilvera Done, going with the create* approach

File: pkg/virtctl/vnc/vnc.go (hunk 2 above)
Comment (maiqueb) [reply pkg/virtctl/vnc/vnc.go:98]: Fair enough.

File: pkg/virtctl/vnc/vnc.go [hunk 5]
```diff
@@ -84,37 +84,51 @@ type VNC struct {
...
+func (o *VNC) Run(cmd *cobra.Command, args []string) error {
+	vmiName := args[0]
+
+	vnc, err := o.createVmiVncStream(vmiName)
 	if err != nil {
-		return fmt.Errorf("Can't listen on unix socket: %s", err.Error())
+		return err
+	}
+
+	ln, err := o.createLocalProxyForVmiWebsockConn(vmiName)
```
Comment (kwiesmueller) [inline pkg/virtctl/vnc/vnc.go:128]: Could you change this to `createLocalProxyForVmiWebsocketConn`.
Websocket is the word used mostly throughout the codebase and if somebody would search for it Websock would not show up.

**Review by kwiesmueller — COMMENTED** [review]
Thanks for looking at this!
I cam across this myself this weekend while working on https://github.com/kubevirt/kubevirt/pull/4944 and the entire subresource and vnc/console part could use some love.

**Comment by enp0s3** [comment]
Addressed comment s from @ezrasilvera and @kwiesmueller. PTAL, thanks!

File: pkg/virtctl/vnc/vnc.go [hunk 6]
```diff
@@ -84,37 +84,51 @@ type VNC struct {
 	clientConfig clientcmd.ClientConfig
 }
 
-func (o *VNC) Run(cmd *cobra.Command, args []string) error {
-	namespace, _, err := o.clientConfig.Namespace()
+func (o *VNC) createLocalProxyForVmiWebsocketConn(vmiName string) (*net.TCPListener, error) {
+	lnAddr, err := net.ResolveTCPAddr("tcp", fmt.Sprintf("127.0.0.1:%d", customPort))
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("Can't resolve the address: %s", err.Error())
```
Comment (EdDev) [inline pkg/virtctl/vnc/vnc.go:90]: Isn't `%v` with `err` the common way to embed the original error text?

File: pkg/virtctl/vnc/vnc.go [hunk 7]
```diff
@@ -84,37 +84,51 @@ type VNC struct {
...
 	}
 
-	vmi := args[0]
+	ln, err := net.ListenTCP("tcp", lnAddr)
+	if err != nil {
+		return nil, fmt.Errorf("Can't listen on unix socket: %s", err.Error())
```
Comment (EdDev) [inline pkg/virtctl/vnc/vnc.go:95]: Same

**Review by EdDev — COMMENTED** [review]
Just two nits

File: pkg/virtctl/vnc/vnc.go (hunk 6 above)
Comment (kwiesmueller) [reply pkg/virtctl/vnc/vnc.go:90]: These days it's `%w` which will wrap the error: https://blog.golang.org/go1.13-errors

Comment (EdDev) [reply pkg/virtctl/vnc/vnc.go:90]: %w is to be used only if one is interested to expose it as an api of the package. I do not think it is a requirement here, especially when the caller has no way to use it as such.

Comment (kwiesmueller) [reply pkg/virtctl/vnc/vnc.go:90]: Right, then it should be `%v`.

**Pull request closed without merging**